app/models/coupon.rb

```diff
@@ -129,18 +129,53 @@ def to_csv
   def can_edit?
     self.redemptions.empty?
   end
+
+  def validate_new_offers(offer_codes)
+    offer_codes = offer_codes.select(&:present?)
+    if offer_codes.blank?
+      self.errors.add(:offer, "code is required")
+    end
+    offer_codes.map do |offer_code|
+      offer = Offer.new(code: offer_code)
+      if !offer.valid?
+        self.errors.add(:offer, "code #{offer_code} is not valid")
+      end
+      offer
```

Comment:
Great question! It's so that this method returns an array of instantiated offers that are later saved [here](https://github.com/Sailias/couponsgem/blob/154b34fd7eee7ca630e245fbfb1dfd37328aeb24/app/controllers/concerns/couponing/crud_controller_methods.rb#L51). The `self.errors(..)` is what is actually checked for validity.

Without returning `offer`, this method would return an array of `nil`'s whether or not the offer is valid. But that does mean it also returns the same array of instantiated objects whether they are valid or not, which...doesn't make sense with what I'd expect either 🤔

Reply:
What do you think about changing it to something like this:
```ruby
offer_codes.map do |offer_code|
  offer = Offer.new(code: offer_code)
  if !offer.valid?
    self.errors.add(:offer, "code #{offer_code} is not valid")
    next
  else
    offer
  end
end
```

Reply:
I think that this should be done via a before_validation or a accepts_nested_attributes_for.

Reply:
because you already have `params[:offers]` it seems like you could do this with `accepts_nested_attributes_for` and let the offer validate itself on create and then display the errors in the form for the `offer`

Reply:
Yeah, this is a good call out. I wasn't able to make good use of `accepts_nested_attributes_for` for this context the way I was hoping -- but, I was still able to remove that `validate_new_offers` method! You’re right. We can let the offer validate itself.

Reply:
Came back with a refactor that does take advantage of `accepts_nested_attributes_for`. Let me know what you think!